src/storage.rs

```diff
@@ -0,0 +1,215 @@
+mod schema;
+mod upgrade;
+
+use std::path::{Path, PathBuf};
+
+use crate::errors::*;
+
+pub struct Storage {
+    pub db_path: PathBuf,
+    pub root_path: PathBuf,
+    conn: rusqlite::Connection,
+}
+
+impl Storage {
+    pub fn create_at(db_path: &Path) -> Result<()> {
+        info!("Creating database at {}", db_path.display());
+        Self::create_or_open(db_path)?;
+        Ok(())
+    }
+
+    pub fn open(db_path: &Path) -> Result<Self> {
+        info!("Opening database at {}", db_path.display());
+        Self::create_or_open(db_path)
+    }
+
+    /// Open a sqlite3 DB file, also creating it if it doesn't already exist.
+    /// Note that the parent directory will NOT be created if it doesn't exist.
+    fn create_or_open(db_path: &Path) -> Result<Self> {
+        let conn = rusqlite::Connection::open(&db_path)
+            .map_err(|_| ErrorKind::DatabaseAccessError(db_path.to_path_buf()))?;
+
+        // Use a canonical path to avoid issues such as #168
+        let db_path = db_path
+            .canonicalize()
+            .map_err(|_| ErrorKind::NoDatabaseFound(db_path.to_path_buf()))?;
+
+        let mut res = Storage {
+            root_path: determine_root_path(&db_path)?,
+            db_path,
+            conn,
+        };
+
+        res.upgrade_database()?;
+
+        Ok(res)
+    }
+
+    pub fn begin_transaction<'a>(&'a mut self) -> Result<Transaction<'a>> {
+        Ok(Transaction {
+            tx: self.conn.transaction()?,
+        })
+    }
+
+    fn upgrade_database(&mut self) -> Result<()> {
+        let mut tx = self.begin_transaction()?;
+
+        upgrade::upgrade(&mut tx)?;
+
+        tx.commit()?;
+        Ok(())
+    }
+}
+
+fn determine_root_path(db_path: &Path) -> Result<PathBuf> {
+    let parent_opt = db_path.parent();
+    let name_opt = parent_opt.map(|p| p.file_name()).flatten();
+    if let Some(dir_name) = name_opt {
+        // If a directory has a name, parent_opt cannot be None
+        let parent = parent_opt.unwrap();
+
+        if dir_name == ".tmsu" {
+            // The unwrap() call should never fail for a canonical path
+            Ok(parent.parent().unwrap().to_path_buf())
+        } else {
+            // Unexpected directory name: return the direct parent
+            // Note that this differs from the Go implementation
+            Ok(parent.to_path_buf())
```

Comment:
If I understood correctly the Go implementation, the fallback would be `/` here. Using the parent directory makes more sense to me.